_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,471 @@
+
+The `run-model` function is responsible for calling the machine learning model, in this case the `inception` function, and uploading the result to S3.
+
+We will use the `python3-http` template for both functions:
+
+```bash
+# Get the python3-flask template from the store
+faas-cli template store pull python3-flask
+
+# Scaffold the functions
+faas-cli new create-batch --lang python3-http-debian
+mv create-batch.yml stack.yml
+faas-cli new run-model --lang python3-http -f stack.yml
+```
+
+Note that we use `python3-http-debian` for the `create-batch` function. [pandas](https://pypi.org/project/pandas/) will be used to process the CSV file. The pandas pip module requires a native build toolchain. It is advisable to use the debian version of the template for native dependencies.
+
+All dependencies have to be put into the `requirements.txt` file.
+
+The `requirements.txt` file for the `create-batch` function:
+
+```
+requests
+pandas
+smart_open[s3]
+```
+
+The handler of the `create-batch` functions:
+
+```python
+import os
+import json
+import uuid
+import requests
+import pandas as pd
+import boto3
+
+from smart_open import open
+
+s3Client = None
+
+def initS3():
+    with open('/var/openfaas/secrets/s3-key', 'r') as s:
+        s3Key = s.read()
+    with open('/var/openfaas/secrets/s3-secret', 'r') as s:
+        s3Secret = s.read()
+
+    session = boto3.Session(
+        aws_access_key_id=s3Key,
+        aws_secret_access_key=s3Secret,
+    )
+    
+    return session.client('s3')
+
+def handle(event, context):
+    global s3Client
+
+    # Initialize an S3 client upon first invocation
+    if s3Client == None:
+        s3Client = initS3()
+    
+    bucketName = os.getenv('s3_bucket')
+
+    # Get the data set from S3
+    batchFile = event.body.decode()
+    s3URL = 's3://{}/{}'.format(bucketName, batchFile)
+    with open(s3URL, 'rb', transport_params={'client': s3Client }) as f:
+        records = pd.read_csv(f)
+
+    batchId = str(uuid.uuid4())
+    batchSize = len(records)
+
+    # Fan-out by invoking the run-model function asynchronously
+    for index, col in records.iterrows():
+        headers = { 'X-Batch-Id': batchId }
+        res = requests.post('http://gateway.openfaas:8080/async-function/run-model', data=col['url'], headers=headers)
+
+    response = {
+        'batch_id': batchId,
+        'batch_size': batchSize
+    }
+
+    return {
+        "statusCode": 201,
+        "body": json.dumps(response)
+    }
+```
+
+At the start of the function the S3 credentials are read and the name of the bucket is passed in using an env variable. Make sure the correct secrets and env variables are added to the `stack.yaml` configuration for each function.
```

Comment:
`At the start of the function the S3 credentials are read and the name of the bucket is passed in using an env variable`
=>
`At the start of the function the S3 credentials are read, and the name of the bucket is passed in using an env variable`